File: Src/Core/Backend/Declare.cpp
/*
* #-----------------------------------------------------# *
 * #                                                     # *
 * #                           Typed                     # *
 * #                   A text formatting DSL             # *
 * #                                                     # *
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #         Created by Rodrigo R. & Contributors        # *
 * #         Released under the Apache License 2.0       # *
 * #            Check LICENSE.MD for more info           # *
 * #                                                     # *
 * #-----------------------------------------------------# *
*/

//
// Created by Rodrigo on 5/21/26.
//

#include "ADT/Exception/ExpectedInitialValue.h"
#include "ADT/Exception/UnexpectedToken.h"
#include "Support/Runtime/AccessString.h"
#include "Support/Runtime/TypeChecker.h"
#include "Walker.h"

using namespace Typed;
using namespace Typed::Core;
using namespace Typed::Core::Backend;

void Walker::Declare(
    VariableStack &stack,
    TreePtr body
)
{
    // Note: redefining variables is allowed in Typed
    auto &name = body->children[0]->value;
    auto &type = body->children[1]->type;
    bool has_init = body->children.size() > 2;

    if (
        type != ADT::Lang::ASTType::List &&
        type != ADT::Lang::ASTType::Dictionary &&
        !has_init
    )
    {
        throw ADT::Exception::ExpectedInitialValue(
            body->line,
            body->column
        );
    }

    if (type == ADT::Lang::ASTType::List && !has_init)
    {
        ADT::Runtime::Object obj;
        obj.type = ADT::Runtime::ObjectType::List;
        obj.value = ADT::List::DynamicObject::Make();
        stack.Emplace(name, std::move(obj));
    } else if (type == ADT::Lang::ASTType::Dictionary && !has_init)
    {
        ADT::Runtime::Object obj;
        obj.type = ADT::Runtime::ObjectType::Dictionary;
        obj.value = ADT::Map::Object::Make();
        stack.Emplace(name, std::move(obj));
    }
    else if (type == ADT::Lang::ASTType::Regex)
    {
        auto &value = body->children[2];
        auto obj = Expression(stack, value);

        // Best case: if the object is already a regex
        if (obj.type == ADT::Runtime::ObjectType::Regex)
        {
            stack.Emplace(name, std::move(obj));
            return;
        }

        Support::Runtime::TypeCheck(
            ADT::Runtime::ObjectType::String,
            obj.type,
            body->line,
            body->column
        );

        ADT::Regex::Ref regex;
        auto str = Support::Runtime::AccessString(obj);

        regex.Build(str);
        stack.Emplace(
            name,
            ADT::Runtime::ObjectType::Regex,
            std::move(regex)
        );
    }
    else
    {
        auto &value = body->children[2];
        auto obj = Expression(stack, value);

        // Do type checking
        Support::Runtime::TypeCheck(
            type == ADT::Lang::ASTType::String ?
                ADT::Runtime::ObjectType::String :
            type == ADT::Lang::ASTType::Integer ?
                ADT::Runtime::ObjectType::Integer :
            type == ADT::Lang::ASTType::Float ?
                ADT::Runtime::ObjectType::Float :
            type == ADT::Lang::ASTType::Boolean ?
                ADT::Runtime::ObjectType::Boolean :
            type == ADT::Lang::ASTType::List ?
                ADT::Runtime::ObjectType::List :
                ADT::Runtime::ObjectType::Dictionary,
            obj.type,
            body->line,
            body->column
        );

        stack.Emplace(name, std::move(obj));
    }
}

File: src/core/backend/declare.rs
/*
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #                           Typed                     # *
 * #                   A text formatting DSL             # *
 * #                                                     # *
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #         Created by Rodrigo R. & Contributors        # *
 * #         Released under the Apache License 2.0       # *
 * #            Check LICENSE.MD for more info           # *
 * #                                                     # *
 * #-----------------------------------------------------# *
*/
use std::cell::{RefCell, RefMut};
use crate::adt::lang::{ASTValue, File, Kind, AST};
use crate::adt::result::ExecutionResult;
use crate::adt::runtime::{*};
use crate::adt::runtime::Object::NonHashable;
use crate::adt::variable::ScopedStack;
use crate::core::backend::expression;
use crate::support::runtime::kind::{check_kind, convert_kind};

pub fn evaluate<'a>(
    file: &'a File<'a>,
    expr: &AST<'a>,
    stack: &mut RefMut<ScopedStack<'a>>
) -> ExecutionResult<'a> {
    let children = expr.children.borrow();

    let data = children.get(0).unwrap();
    let data = data.borrow();
    let (name, kind) = convert_kind(&data);
    let initial;

    // Determine if the declaration has an initializer
    if children.len() > 1 {
        let init = children.get(1).unwrap().borrow();
        initial = expression::evaluate(file, &init, stack)?.0;

        check_kind(
            kind,
            &initial,
            expr
        )?;

    } else {
        match kind {
            Kind::Integer => {
                initial = Object::Hashable(HashableObject::Integer(0));
            }

            Kind::Float => {
                initial = Object::Hashable(
                    HashableObject::Float(
                        Float::from(0.0)
                    )
                );
            }

            Kind::String => {
                initial = Object::Hashable(
                    HashableObject::String(
                        ASTValue::Borrowed("")
                    )
                );
            }

            Kind::Boolean => {
                initial = Object::Hashable(
                    HashableObject::Boolean(false)
                );
            }

            Kind::Dictionary => {
                initial = NonHashable(
                    NonHashableObject::Dictionary(
                        RuntimeDictionary::new(RefCell::new(Dictionary::new()))
                    )
                );
            }

            Kind::List => {
                initial = NonHashable(
                    NonHashableObject::List(
                        RuntimeList::new(RefCell::new(List::new()))
                    )
                );
            }
            
            _ => unreachable!()
        }
    }

    // Insert the variable into the stack
    stack.push(name, &initial);

    Ok(Object::Void)
}